File: src/pattern.rs
use super::types::*;

impl Pattern {
    pub fn match_(&self, term: &Value) -> Option<Vec<Value>> {
        match (self, term) {
            (Pattern::EffectPure(_), Value::RequestWithContinuation(_, _, _, _, _)) => None,
            (Pattern::EffectPure(pattern), Value::RequestPure(inner)) => pattern.match_(inner),
            (
                Pattern::EffectBind(reference, number, args, kont),
                Value::RequestWithContinuation(tref, tnum, targs, tidx, tkont),
            ) if reference == tref && number == tnum && args.len() == targs.len() => {
                let mut all = vec![];
                for i in 0..args.len() {
                    match args[i].match_(&targs[i]) {
                        None => return None,
                        Some(inner) => {
                            all.extend(inner);
                        }
                    }
                }
                match &**kont {
                    Pattern::Unbound => (),
                    Pattern::Var => all.push(Value::Continuation(*tidx, tkont.clone())),
                    _ => unreachable!("Can't match on a continuation"),
                }
                Some(all)
            }
            (Pattern::Unbound, _) => Some(vec![]),
            (Pattern::Var, t) => Some(vec![t.clone()]),
            (Pattern::Boolean(a), Value::Boolean(b)) if a == b => Some(vec![]),
            (Pattern::Int(a), Value::Int(b)) if a == b => Some(vec![]),
            (Pattern::Nat(a), Value::Nat(b)) if a == b => Some(vec![]),
            (Pattern::Float(a), Value::Float(b)) if a == b => Some(vec![]),
            (Pattern::Text(a), Value::Text(b)) if a == b => Some(vec![]),
            (Pattern::Char(a), Value::Char(b)) if a == b => Some(vec![]),
            (Pattern::As(a), term) => match a.match_(term) {
                None => None,
                Some(mut terms) => {
                    terms.insert(0, term.clone());
                    Some(terms)
                }
            },
            (Pattern::SequenceLiteral(patterns), Value::Sequence(items))
                if patterns.len() == items.len() =>
            {
                let mut all = vec![];
                for i in 0..patterns.len() {
                    match patterns[i].match_(&items[i]) {
                        None => return None,
                        Some(v) => {
                            all.extend(v);
                        }
                    }
                }
                return Some(all);
            }
            (Pattern::SequenceOp(one, op, two), Value::Sequence(contents)) => match op {
                SeqOp::Cons => {
                    if contents.len() > 0 {
                        match one.match_(&contents[0]) {
                            None => None,
                            Some(mut ones) => {
                                match two.match_(&Value::Sequence(contents[1..].to_vec())) {
                                    None => None,
                                    Some(twos) => {
                                        ones.extend(twos);
                                        Some(ones)
                                    }
                                }
                            }
                        }
                    } else {
                        None
                    }
                }
                SeqOp::Snoc => {
                    if contents.len() > 0 {
                        match one.match_(&Value::Sequence(contents[..contents.len() - 1].to_vec()))
                        {
                            None => None,
                            Some(mut ones) => match two.match_(&contents[contents.len() - 1]) {
                                None => None,
                                Some(twos) => {
                                    ones.extend(twos);
                                    Some(ones)
                                }
                            },
                        }
                    } else {
                        None
                    }
                }
                SeqOp::Concat => match (&**one, &**two) {
                    (Pattern::SequenceLiteral(patterns), two) => {
                        if contents.len() >= patterns.len() {
                            match one.match_(&Value::Sequence(contents[0..patterns.len()].to_vec()))
                            {
                                None => None,
                                Some(mut ones) => {
                                    match two.match_(&Value::Sequence(
                                        contents[patterns.len()..].to_vec(),
                                    )) {
                                        None => None,
                                        Some(twos) => {
                                            ones.extend(twos);
                                            Some(ones)
                                        }
                                    }
                                }
                            }
                        } else {
                            None
                        }
                    }
                    (_, Pattern::SequenceLiteral(patterns)) => {
                        if contents.len() >= patterns.len() {
                            let split = contents.len() - patterns.len();
                            match one.match_(&Value::Sequence(contents[0..split].to_vec())) {
                                None => None,
                                Some(mut ones) => {
                                    match two.match_(&Value::Sequence(contents[split..].to_vec())) {
                                        None => None,
                                        Some(twos) => {
                                            ones.extend(twos);
                                            Some(ones)
                                        }
                                    }
                                }
                            }
                        } else {
                            None
                        }
                    }
                    _ => unreachable!(),
                },
            },
            (Pattern::Constructor(reference, number, children), inner) => {
                let mut all = vec![];
                if children.len() > 0 {
                    match inner {
                        Value::PartialConstructor(r, n, pchildren)
                            if r == reference && n == number =>
                        {
                            if pchildren.len() != children.len() {
                                return None;
                            }
                            for i in 0..children.len() {
                                match children[i].match_(&pchildren[i]) {
                                    None => return None,
                                    Some(v) => {
                                        all.extend(v);
                                    }
                                }
                            }
                            return Some(all);
                        }
                        _ => return None,
                    }
                }

                match inner {
                    Value::Constructor(r, n) if r == reference && n == n => Some(all),
                    _ => None,
                }
                //
            }
            _ => None,
        }
    }
}

File: js/pattern.js
// ok

import clone from 'clone-deep';
import compare from './compare';
const key = (o) => (typeof o === 'string' ? o : Object.keys(o)[0]);

export const patternMatch = (pattern, term) => {
    const pt = key(pattern);
    const vt = key(term);
    if (pt === 'Unbound') {
        return [];
    }
    if (pt === 'Var') {
        return [term];
    }
    if (pt === 'As') {
        const res = patternMatch(pattern[pt], term);
        return res != null ? [term].concat(res) : null;
    }
    if (pt === 'Constructor') {
        return matchConstructor(pattern[pt], term);
    }
    // Boolean, Int, Nat, Float, Text, Char
    if (pt === vt) {
        // console.log('builtin', pattern[pt], term[vt], pt);
        return pattern[pt] === term[vt] ? [] : null;
    }
    const bothKey = `${pt}:${vt}`;
    // console.log('bk', bothKey);
    if (matchers[bothKey]) {
        return matchers[bothKey](pattern[pt], term[vt]);
    }
    return null;
};

const matchers = {
    'EffectPure:RequestPure': (pattern, value) => patternMatch(pattern, value),
    'EffectBind:RequestWithContinuation': (
        [reference, number, args, kont],
        [tref, tnum, targs, tidx, tkont, current_idx],
    ) => {
        if (
            compare(reference, tref) === 0 &&
            number === tnum &&
            args.length === targs.length
        ) {
            const all = [];
            for (let i = 0; i < args.length; i++) {
                const res = patternMatch(args[i], targs[i]);
                if (res != null) {
                    all.push(...res);
                } else {
                    return null;
                }
            }
            const kk = key(kont);
            if (kk === 'Var') {
                tkont = tkont.slice(0, current_idx + 1).map((t) => ({ ...t }));
                tkont[current_idx].handler = null;
                all.push({ Continuation: [tidx, tkont] });
            } else if (kk !== 'Unbound') {
                throw new Error('Unable to match on a continuation');
            }
            return all;
        } else {
            return null;
        }
    },
    'SequenceLiteral:Sequence': (patterns, items) => {
        if (patterns.length === items.length) {
            const all = [];
            const failed = patterns.some((p, i) => {
                const res = patternMatch(p, items[i]);
                if (res != null) {
                    all.push(...res);
                } else {
                    return true;
                }
            });
            if (failed) {
                return null;
            }
            return all;
        }
    },
    'SequenceOp:Sequence': ([one, op, two], contents) => {
        if (op === 'Cons') {
            if (contents.length > 0) {
                const res = patternMatch(one, contents[0]);
                if (res != null) {
                    const r2 = patternMatch(two, contents.slice(1));
                    if (r2 != null) {
                        res.concat(...r2);
                        return res;
                    }
                }
            }
            return null;
        } else if (op === 'Snoc') {
            if (contents.length > 0) {
                const res = patternMatch(one, contents.slice(0, -1));
                if (res != null) {
                    const r2 = patternMatch(two, contents[contents.length - 1]);
                    if (r2 != null) {
                        res.concat(...r2);
                        return res;
                    }
                }
            }
            return null;
        } else if (op === 'Concat') {
            const ko = key(one);
            const kt = key(two);
            if (ko === 'SequenceLiteral') {
                const patterns = one[ko];
                if (contents.length >= patterns.length) {
                    const res = patternMatch(one, {
                        Sequence: contents.slice(0, patterns.length),
                    });
                    if (res != null) {
                        const r2 = patternMatch(two, {
                            Sequence: contents.slice(patterns.length),
                        });
                        if (r2 != null) {
                            res.push(...r2);
                            return res;
                        }
                    }
                }
                return null;
            }
            if (kt === 'SequenceLiteral') {
                const patterns = two[kt];
                if (contents.length >= patterns.length) {
                    const split = contents.length - patterns.length;
                    const res = patternMatch(one, {
                        Sequence: contents.slice(0, split),
                    });
                    if (res != null) {
                        const r2 = patternMatch(two, {
                            Sequence: contents.slice(split),
                        });
                        if (r2 != null) {
                            res.push(...r2);
                            return res;
                        }
                    }
                }
                return null;
            }
            throw new Error('Concat needs sequence literal');
        }
    },
};

const matchConstructor = ([reference, number, children], inner) => {
    const all = [];
    const k = key(inner);
    if (children.length > 0) {
        if (k === 'PartialConstructor') {
            const [r, n, pchildren] = inner[k];
            if (compare(r, reference) == 0 && number === n) {
                if (pchildren.length !== children.length) {
                    // console.log('children len diff');
                    return null;
                }
                for (let i = 0; i < children.length; i++) {
                    const res = patternMatch(children[i], pchildren[i]);
                    if (res != null) {
                        // console.log('got it', res);
                        all.push(...res);
                    } else {
                        return null;
                    }
                }
                return all;
            } else {
                // console.log('partial not', r, reference, number, n);
            }
        }
    } else {
        if (k === 'Constructor') {
            const [r, nn] = inner[k];
            if (compare(r, reference) == 0 && number === nn) {
                return all;
            }
        }
    }
    return null;
};
